src/c/c_client.cpp

```diff
@@ -1674,3 +1674,389 @@ SRError use_tensor_ensemble_prefix(void* c_client, bool use_prefix)
+    *datasets = NULL;
+    if (*num_datasets > 0) {
+      *datasets = new void*[*num_datasets];
+      for (size_t i = 0; i < *num_datasets; i++) {
+        datasets[i] = (void*)(new DataSet(std::move(result_datasets[i])));
+      }
+    }
+  }
+  catch (const Exception& e) {
+    SRSetLastError(e);
+    result = e.to_error_code();
+  }
+  catch (...) {
+    SRSetLastError(SRInternalException("Unknown exception occurred"));
+    result = SRInternalError;
+  }
+
+  return result;
+}
+
+// Get a range of datasets (by index) from an aggregation list
+extern "C"
+SRError get_dataset_list_range(void* c_client, const char* list_name,
+                               const size_t list_name_length,
+                               const int start_index, const int end_index,
+                               void** datasets, size_t* num_datasets)
+{
+  SRError result = SRNoError;
+  try
+  {
+    // Sanity check params
+    SR_CHECK_PARAMS(c_client != NULL && list_name != NULL &&
+                    datasets != NULL && num_datasets != NULL);
+
+    Client* s = reinterpret_cast<Client*>(c_client);
+    std::string lname(list_name, list_name_length);
+
+    std::vector<DataSet> result_datasets = s->get_dataset_list_range(
+      lname, start_index, end_index);
+    *num_datasets = result_datasets.size();
+    *datasets = NULL;
+    if (*num_datasets > 0) {
+      *datasets = new void*[*num_datasets];
+      for (size_t i = 0; i < *num_datasets; i++) {
+        datasets[i] = (void*)(new DataSet(std::move(result_datasets[i])));
+      }
+    }
+  }
+  catch (const Exception& e) {
+    SRSetLastError(e);
+    result = e.to_error_code();
+  }
+  catch (...) {
+    SRSetLastError(SRInternalException("Unknown exception occurred"));
+    result = SRInternalError;
+  }
+
+  return result;
+}
+
+// Get a range of datasets (by index) from an aggregation list into an
+// already allocated vector of datasets
+extern "C"
+SRError _get_dataset_list_range_allocated(void* c_client, const char* list_name,
+                               const size_t list_name_length,
+                               const int start_index, const int end_index,
+                               void** datasets)
+{
+  SRError result = SRNoError;
+  try
+  {
+    // Sanity check params
+    SR_CHECK_PARAMS(c_client != NULL && list_name != NULL &&
+                    datasets != NULL);
+
+    Client* s = reinterpret_cast<Client*>(c_client);
+    std::string lname(list_name, list_name_length);
+
+    std::vector<DataSet> result_datasets = s->get_dataset_list_range(
+      lname, start_index, end_index);
+    int num_datasets = result_datasets.size();
+    if ( num_datasets != (end_index-start_index+1)) {
+      SRSetLastError(SRInternalException(
+        "Returned dataset list is not equal to the requested range"
+      ));
+    }
+
+    if (num_datasets > 0) {
+      for (size_t i = 0; i < num_datasets; i++) {
+        datasets[i] = (void*)(new DataSet(std::move(result_datasets[i])));
+      }
```

Comment:
If this routine doesn't work, change the declaration to a void*** and make this line:
`(*datasets)[i] = (void*)(new DataSet(std::move(result_datasets[i])));`

Reply:
This worked fine for the Fortran implementation which is somewhat baffling to me. I think perhaps because the memory is allocated on the Fortran side, and then we're passing a `c_ptr` (essentially a `void *`) to the Fortran array. Since Fortran is passing everything by reference, then it is correctly pointing to the first element of the Fortran array after which C is just striding through memory at the size of a `void *`.